src/backend/allocator/vulkan/mod.rs

```diff
@@ -0,0 +1,770 @@
+        };
+
+        let (dropped_sender, dropped_recv) = mpsc::channel();
+
+        let mut allocator = VulkanAllocator {
+            formats: Vec::new(),
+            images: Vec::new(),
+            remaining_allocations: phd.limits().max_memory_allocation_count,
+            extension_fns,
+            dropped_recv,
+            dropped_sender,
+            phd: phd.clone(),
+            device: Arc::new(device),
+        };
+
+        allocator.init_formats();
+
+        Ok(allocator)
+    }
+
+    /// Returns whether this allocator supports the specified format with the usage flags.
+    pub fn is_format_supported(&self, format: DrmFormat, usage: ImageUsageFlags) -> bool {
+        // TODO: Check if the extents are also valid?
+        // Vulkan states a maximum extent size for images.
+        // This may also be useful as a function on Allocator.
+        unsafe { self.get_format_info(format, vk::ImageUsageFlags::from_raw(usage.bits())) }
+            .ok()
+            .is_some()
+    }
+
+    /// Try to create a buffer with the given dimensions, pixel format and usage flags.
+    pub fn create_buffer_with_usage(
+        &mut self,
+        width: u32,
+        height: u32,
+        fourcc: DrmFourcc,
+        modifiers: &[DrmModifier],
+        usage: ImageUsageFlags,
+    ) -> Result<VulkanImage, Error> {
+        self.cleanup();
+
+        let vk_format = format::get_vk_format(fourcc).ok_or(Error::UnsupportedFormat)?;
+        let vk_usage = vk::ImageUsageFlags::from_raw(usage.bits());
+
+        // VUID-VkImageCreateInfo-extent-00944, VUID-VkImageCreateInfo-extent-00945
+        if width == 0 || height == 0 {
+            return Err(Error::InvalidSize);
+        }
+
+        // Filter out any format + modifier combinations that are not supported
+        let modifiers = self.filter_modifiers(width, height, vk_usage, fourcc, modifiers);
+
+        // VUID-VkImageDrmFormatModifierListCreateInfoEXT-drmFormatModifierCount-arraylength
+        if modifiers.is_empty() {
+            return Err(Error::UnsupportedFormat);
+        }
+
+        Ok(unsafe { self.create_image(width, height, vk_format, vk_usage, fourcc, &modifiers[..]) }?)
+    }
+
+    /// Returns the [`PhysicalDevice`] this allocator was created with.
+    pub fn physical_device(&self) -> &PhysicalDevice {
+        &self.phd
+    }
+}
+
+impl Allocator<VulkanImage> for VulkanAllocator {
+    type Error = Error;
+
+    fn create_buffer(
+        &mut self,
+        width: u32,
+        height: u32,
+        fourcc: DrmFourcc,
+        modifiers: &[DrmModifier],
+    ) -> Result<VulkanImage, Self::Error> {
+        self.create_buffer_with_usage(
+            width,
+            height,
+            fourcc,
+            modifiers,
+            // More than likely images from the allocator will be sampled or used with color attachments when
+            // exported.
+            ImageUsageFlags::SAMPLED | ImageUsageFlags::COLOR_ATTACHMENT,
```

Comment:
> Do we need to account for what other implementation might do to the buffer after importing them?

Yes.

The relevant section of the Vulkan specification (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/man/html/vkGetPhysicalDeviceImageFormatProperties.html) for determining if an image format is supported states the following:

> If format is not a supported image format, or if the combination of format, type, tiling, usage, and flags is not supported for images, then vkGetPhysicalDeviceImageFormatProperties returns VK_ERROR_FORMAT_NOT_SUPPORTED.

Now wait aren't dmabufs just memory we assign meaning to? Yes.

In Vulkan we rely on the image handle we create to provide us information about what type of memory allocation is needed and then bind the memory to the image. With the drm tiling extension we use in the allocator, the drm tiling could of course change what type of memory the implementation wants to allocate.

Declaring this ahead of time is important for the dmabuf other graphics apis may import as well. Clients and the compositor if they run Vulkan may use the same function above.

If the compositor needs more restrictive usage flags than clients (since it uses client images as textures), then the following applies:
> The limitations on an image format that are reported by vkGetPhysicalDeviceImageFormatProperties have the following property: if usage1 and usage2 of type [VkImageUsageFlags](https://www.khronos.org/registry/vulkan/specs/1.3-extensions/man/html/VkImageUsageFlags.html) are such that the bits set in usage1 are a subset of the bits set in usage2, and flags1 and flags2 of type [VkImageCreateFlags](https://www.khronos.org/registry/vulkan/specs/1.3-extensions/man/html/VkImageCreateFlags.html) are such that the bits set in flags1 are a subset of the bits set in flags2, then the limitations for usage1 and flags1 must be no more strict than the limitations for usage2 and flags2, for all values of format, type, and tiling.

TLDR: Allocating an image with less restrictive usage flags may result in the image not being importable with more restrictive usages.

You do have a point with specifying the default usage flags during construction, which may allow the implementation to request the least restrictive usage flags it needs (which would possibly provide more usable formats).